# Keras Exercise

## Predict political party based on votes

As a fun little example, we'll use a public data set of how US congressmen voted on 17 different issues in the year 1984. Let's see if we can figure out their political party based on their votes alone, using a deep neural network!

For those outside the United States, our two main political parties are "Democrat" and "Republican." In modern times they represent progressive and conservative ideologies, respectively.

Politics in 1984 weren't quite as polarized as they are today, but you should still be able to get over 90% accuracy without much trouble.

Since the point of this exercise is implementing neural networks in Keras, I'll help you to load and prepare the data.

Let's start by importing the raw CSV file using Pandas, and make a DataFrame out of it with nice column labels:

In [4]:
import pandas as pd

feature_names =  ['party','handicapped-infants', 'water-project-cost-sharing', 
                    'adoption-of-the-budget-resolution', 'physician-fee-freeze',
                    'el-salvador-aid', 'religious-groups-in-schools',
                    'anti-satellite-test-ban', 'aid-to-nicaraguan-contras',
                    'mx-missle', 'immigration', 'synfuels-corporation-cutback',
                    'education-spending', 'superfund-right-to-sue', 'crime',
                    'duty-free-exports', 'export-administration-act-south-africa']

voting_data = pd.read_csv('house-votes-84.data.txt', na_values=['?'], 
                          names = feature_names)
voting_data.head()

party handicapped-infants water-project-cost-sharing  \
0  republican                   n                          y   
1  republican                   n                          y   
2    democrat                 NaN                          y   
3    democrat                   n                          y   
4    democrat                   y                          y   

  adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
0                                 n                    y               y   
1                                 n                    y               y   
2                                 y                  NaN               y   
3                                 y                    n             NaN   
4                                 y                    n               y   

  religious-groups-in-schools anti-satellite-test-ban  \
0                           y                       n   
1                           y                       n   
2                           y                       n   
3                           y                       n   
4                           y                       n   

  aid-to-nicaraguan-contras mx-missle immigration  \
0                         n         n           y   
1                         n         n           n   
2                         n         n           n   
3                         n         n           n   
4                         n         n           n   

  synfuels-corporation-cutback education-spending superfund-right-to-sue  \
0                          NaN                  y                      y   
1                            n                  y                      y   
2                            y                  n                      y   
3                            y                  n                      y   
4                            y                NaN                      y   

  crime duty-free-exports export-administration-act-south-africa  
0     y                 n                                      y  
1     y                 n                                    NaN  
2     y                 n                                      n  
3     n                 n                                      y  
4     y                 y                                      y

We can use describe() to get a feel of how the data looks in aggregate:

In [6]:
voting_data.describe() #nature of the data overview

party handicapped-infants water-project-cost-sharing  \
count        435                 423                        387   
unique         2                   2                          2   
top     democrat                   n                          y   
freq         267                 236                        195   

       adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
count                                424                  424             420   
unique                                 2                    2               2   
top                                    y                    n               y   
freq                                 253                  247             212   

       religious-groups-in-schools anti-satellite-test-ban  \
count                          424                     421   
unique                           2                       2   
top                              y                       y   
freq                           272                     239   

       aid-to-nicaraguan-contras mx-missle immigration  \
count                        420       413         428   
unique                         2         2           2   
top                            y         y           y   
freq                         242       207         216   

       synfuels-corporation-cutback education-spending superfund-right-to-sue  \
count                           414                404                    410   
unique                            2                  2                      2   
top                               n                  n                      y   
freq                            264                233                    209   

       crime duty-free-exports export-administration-act-south-africa  
count    418               407                                    331  
unique     2                 2                                      2  
top        y                 n                                      y  
freq     248               233                                    269

We can see there's some missing data to deal with here; some politicians abstained on some votes, or just weren't present when the vote was taken. We will just drop the rows with missing data to keep it simple, but in practice you'd want to first make sure that doing so didn't introduce any sort of bias into your analysis (if one party abstains more than another, that could be problematic for example.)

In [7]:
voting_data.dropna(inplace=True)#drop rows with missing data
voting_data.describe()

party handicapped-infants water-project-cost-sharing  \
count        232                 232                        232   
unique         2                   2                          2   
top     democrat                   n                          n   
freq         124                 136                        125   

       adoption-of-the-budget-resolution physician-fee-freeze el-salvador-aid  \
count                                232                  232             232   
unique                                 2                    2               2   
top                                    y                    n               y   
freq                                 123                  119             128   

       religious-groups-in-schools anti-satellite-test-ban  \
count                          232                     232   
unique                           2                       2   
top                              y                       y   
freq                           149                     124   

       aid-to-nicaraguan-contras mx-missle immigration  \
count                        232       232         232   
unique                         2         2           2   
top                            y         n           y   
freq                         119       119         128   

       synfuels-corporation-cutback education-spending superfund-right-to-sue  \
count                           232                232                    232   
unique                            2                  2                      2   
top                               n                  n                      y   
freq                            152                124                    127   

       crime duty-free-exports export-administration-act-south-africa  
count    232               232                                    232  
unique     2                 2                                      2  
top        y                 n                                      y  
freq     149               146                                    189

Our neural network needs normalized numbers, not strings, to work. So let's replace all the y's and n's with 1's and 0's, and represent the parties as 1's and 0's as well.

In [11]:
voting_data.replace(('y', 'n'), (1, 0), inplace=True) #replace y and n with 1 and 0
voting_data.replace(('democrat', 'republican'), (1, 0), inplace=True)#same here

In [13]:
voting_data.head()

party  handicapped-infants  water-project-cost-sharing  \
5       1                    0                           1   
8       0                    0                           1   
19      1                    1                           1   
23      1                    1                           1   
25      1                    1                           0   

    adoption-of-the-budget-resolution  physician-fee-freeze  el-salvador-aid  \
5                                   1                     0                1   
8                                   0                     1                1   
19                                  1                     0                0   
23                                  1                     0                0   
25                                  1                     0                0   

    religious-groups-in-schools  anti-satellite-test-ban  \
5                             1                        0   
8                             1                        0   
19                            0                        1   
23                            0                        1   
25                            0                        1   

    aid-to-nicaraguan-contras  mx-missle  immigration  \
5                           0          0            0   
8                           0          0            0   
19                          1          1            0   
23                          1          1            0   
25                          1          1            1   

    synfuels-corporation-cutback  education-spending  superfund-right-to-sue  \
5                              0                   0                       1   
8                              0                   1                       1   
19                             1                   0                       0   
23                             0                   0                       0   
25                             0                   0                       0   

    crime  duty-free-exports  export-administration-act-south-africa  
5       1                  1                                       1  
8       1                  0                                       1  
19      0                  1                                       1  
23      0                  1                                       1  
25      0                  1                                       1

Finally let's extract the features and labels in the form that Keras will expect:

In [15]:
all_features = voting_data[feature_names].drop('party', axis=1).values #extract all the feature columns
all_classes = voting_data['party'].values #extract the party columns

OK, so have a go at it! You'll want to refer back to the slide on using Keras with binary classification - there are only two parties, so this is a binary problem. This also saves us the hassle of representing classes with "one-hot" format like we had to do with MNIST; our output is just a single 0 or 1 value.

Also refer to the scikit_learn integration slide, and use cross_val_score to evaluate your resulting model with 10-fold cross-validation.

**If you're using tensorflow-gpu on a Windows machine** by the way, you probably *do* want to peek a little bit at my solution - if you run into memory allocation errors, there's a workaround there you can use.

Try out your code here:

In [40]:
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.models import Sequential
from sklearn.model_selection import cross_val_score
def create_model():
    model = Sequential()
    model.add(Dense(64, input_dim = 20, activation = 'relu', ))
    model.add(Dropout(0.5))
    model.add(Dense(64, activation = 'relu', ))
    model.add(Dropout(0.5))
    model.add(Dense(1, activation = 'sigmoid'))
    model.compile(loss = 'binary_crossentropy', optimizer = 'rmsprop', metrics=['accuracy'])
    return model
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier
estimator = KerasClassifier(build_fn=create_model, epochs = 100, verbose=0)
cv_scores=cross_val_score(estimator, all_features, all_classes, cv=10)

/var/folders/f2/gq8xgm9x491cqwtct_tgpy0h0000gn/T/ipykernel_84668/4147869179.py:14: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  estimator = KerasClassifier(build_fn=create_model, epochs = 100, verbose=0)
2022-09-23 00:19:05.879781: W tensorflow/core/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz


ValueError: 
All the 10 fits failed.
It is very likely that your model is misconfigured.
You can try to debug the error by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 248, in fit
    return super().fit(x, y, **kwargs)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 175, in fit
    history = self.model.fit(x, y, **fit_args)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/var/folders/f2/gq8xgm9x491cqwtct_tgpy0h0000gn/T/__autograph_generated_filekk87zm32.py", line 15, in tf__train_function
    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)
ValueError: in user code:

    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1160, in train_function  *
        return step_function(self, iterator)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1146, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1135, in run_step  **
        outputs = model.train_step(data)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 993, in train_step
        y_pred = self(x, training=True)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
        raise e.with_traceback(filtered_tb) from None
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/input_spec.py", line 295, in assert_input_compatibility
        raise ValueError(

    ValueError: Input 0 of layer "sequential_11" is incompatible with the layer: expected shape=(None, 20), found shape=(None, 16)


--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 248, in fit
    return super().fit(x, y, **kwargs)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 175, in fit
    history = self.model.fit(x, y, **fit_args)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/var/folders/f2/gq8xgm9x491cqwtct_tgpy0h0000gn/T/__autograph_generated_filekk87zm32.py", line 15, in tf__train_function
    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)
ValueError: in user code:

    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1160, in train_function  *
        return step_function(self, iterator)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1146, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1135, in run_step  **
        outputs = model.train_step(data)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 993, in train_step
        y_pred = self(x, training=True)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
        raise e.with_traceback(filtered_tb) from None
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/input_spec.py", line 295, in assert_input_compatibility
        raise ValueError(

    ValueError: Input 0 of layer "sequential_12" is incompatible with the layer: expected shape=(None, 20), found shape=(None, 16)


--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 248, in fit
    return super().fit(x, y, **kwargs)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 175, in fit
    history = self.model.fit(x, y, **fit_args)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/var/folders/f2/gq8xgm9x491cqwtct_tgpy0h0000gn/T/__autograph_generated_filekk87zm32.py", line 15, in tf__train_function
    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)
ValueError: in user code:

    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1160, in train_function  *
        return step_function(self, iterator)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1146, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1135, in run_step  **
        outputs = model.train_step(data)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 993, in train_step
        y_pred = self(x, training=True)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
        raise e.with_traceback(filtered_tb) from None
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/input_spec.py", line 295, in assert_input_compatibility
        raise ValueError(

    ValueError: Input 0 of layer "sequential_13" is incompatible with the layer: expected shape=(None, 20), found shape=(None, 16)


--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 248, in fit
    return super().fit(x, y, **kwargs)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 175, in fit
    history = self.model.fit(x, y, **fit_args)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/var/folders/f2/gq8xgm9x491cqwtct_tgpy0h0000gn/T/__autograph_generated_filekk87zm32.py", line 15, in tf__train_function
    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)
ValueError: in user code:

    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1160, in train_function  *
        return step_function(self, iterator)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1146, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1135, in run_step  **
        outputs = model.train_step(data)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 993, in train_step
        y_pred = self(x, training=True)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
        raise e.with_traceback(filtered_tb) from None
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/input_spec.py", line 295, in assert_input_compatibility
        raise ValueError(

    ValueError: Input 0 of layer "sequential_14" is incompatible with the layer: expected shape=(None, 20), found shape=(None, 16)


--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 248, in fit
    return super().fit(x, y, **kwargs)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 175, in fit
    history = self.model.fit(x, y, **fit_args)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/var/folders/f2/gq8xgm9x491cqwtct_tgpy0h0000gn/T/__autograph_generated_filekk87zm32.py", line 15, in tf__train_function
    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)
ValueError: in user code:

    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1160, in train_function  *
        return step_function(self, iterator)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1146, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1135, in run_step  **
        outputs = model.train_step(data)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 993, in train_step
        y_pred = self(x, training=True)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
        raise e.with_traceback(filtered_tb) from None
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/input_spec.py", line 295, in assert_input_compatibility
        raise ValueError(

    ValueError: Input 0 of layer "sequential_15" is incompatible with the layer: expected shape=(None, 20), found shape=(None, 16)


--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 248, in fit
    return super().fit(x, y, **kwargs)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 175, in fit
    history = self.model.fit(x, y, **fit_args)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/var/folders/f2/gq8xgm9x491cqwtct_tgpy0h0000gn/T/__autograph_generated_filekk87zm32.py", line 15, in tf__train_function
    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)
ValueError: in user code:

    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1160, in train_function  *
        return step_function(self, iterator)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1146, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1135, in run_step  **
        outputs = model.train_step(data)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 993, in train_step
        y_pred = self(x, training=True)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
        raise e.with_traceback(filtered_tb) from None
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/input_spec.py", line 295, in assert_input_compatibility
        raise ValueError(

    ValueError: Input 0 of layer "sequential_16" is incompatible with the layer: expected shape=(None, 20), found shape=(None, 16)


--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 248, in fit
    return super().fit(x, y, **kwargs)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 175, in fit
    history = self.model.fit(x, y, **fit_args)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/var/folders/f2/gq8xgm9x491cqwtct_tgpy0h0000gn/T/__autograph_generated_filekk87zm32.py", line 15, in tf__train_function
    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)
ValueError: in user code:

    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1160, in train_function  *
        return step_function(self, iterator)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1146, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1135, in run_step  **
        outputs = model.train_step(data)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 993, in train_step
        y_pred = self(x, training=True)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
        raise e.with_traceback(filtered_tb) from None
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/input_spec.py", line 295, in assert_input_compatibility
        raise ValueError(

    ValueError: Input 0 of layer "sequential_17" is incompatible with the layer: expected shape=(None, 20), found shape=(None, 16)


--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 248, in fit
    return super().fit(x, y, **kwargs)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 175, in fit
    history = self.model.fit(x, y, **fit_args)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/var/folders/f2/gq8xgm9x491cqwtct_tgpy0h0000gn/T/__autograph_generated_filekk87zm32.py", line 15, in tf__train_function
    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)
ValueError: in user code:

    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1160, in train_function  *
        return step_function(self, iterator)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1146, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1135, in run_step  **
        outputs = model.train_step(data)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 993, in train_step
        y_pred = self(x, training=True)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
        raise e.with_traceback(filtered_tb) from None
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/input_spec.py", line 295, in assert_input_compatibility
        raise ValueError(

    ValueError: Input 0 of layer "sequential_18" is incompatible with the layer: expected shape=(None, 20), found shape=(None, 16)


--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 248, in fit
    return super().fit(x, y, **kwargs)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 175, in fit
    history = self.model.fit(x, y, **fit_args)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/var/folders/f2/gq8xgm9x491cqwtct_tgpy0h0000gn/T/__autograph_generated_filekk87zm32.py", line 15, in tf__train_function
    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)
ValueError: in user code:

    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1160, in train_function  *
        return step_function(self, iterator)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1146, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1135, in run_step  **
        outputs = model.train_step(data)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 993, in train_step
        y_pred = self(x, training=True)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
        raise e.with_traceback(filtered_tb) from None
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/input_spec.py", line 295, in assert_input_compatibility
        raise ValueError(

    ValueError: Input 0 of layer "sequential_19" is incompatible with the layer: expected shape=(None, 20), found shape=(None, 16)


--------------------------------------------------------------------------------
1 fits failed with the following error:
Traceback (most recent call last):
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/sklearn/model_selection/_validation.py", line 686, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 248, in fit
    return super().fit(x, y, **kwargs)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/wrappers/scikit_learn.py", line 175, in fit
    history = self.model.fit(x, y, **fit_args)
  File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
    raise e.with_traceback(filtered_tb) from None
  File "/var/folders/f2/gq8xgm9x491cqwtct_tgpy0h0000gn/T/__autograph_generated_filekk87zm32.py", line 15, in tf__train_function
    retval_ = ag__.converted_call(ag__.ld(step_function), (ag__.ld(self), ag__.ld(iterator)), None, fscope)
ValueError: in user code:

    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1160, in train_function  *
        return step_function(self, iterator)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1146, in step_function  **
        outputs = model.distribute_strategy.run(run_step, args=(data,))
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 1135, in run_step  **
        outputs = model.train_step(data)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/training.py", line 993, in train_step
        y_pred = self(x, training=True)
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/utils/traceback_utils.py", line 70, in error_handler
        raise e.with_traceback(filtered_tb) from None
    File "/Users/antonis/miniforge3/envs/mlp/lib/python3.8/site-packages/keras/engine/input_spec.py", line 295, in assert_input_compatibility
        raise ValueError(

    ValueError: Input 0 of layer "sequential_20" is incompatible with the layer: expected shape=(None, 20), found shape=(None, 16)



## My implementation is below

# No peeking!

![title](peek.jpg)

In [7]:
from tensorflow.keras.layers import Dense, Dropout
from tensorflow.keras.models import Sequential
from sklearn.model_selection import cross_val_score

def create_model():
    model = Sequential()
    #16 feature inputs (votes) going into an 32-unit layer 
    model.add(Dense(32, input_dim=16, kernel_initializer='normal', activation='relu'))
    # Another hidden layer of 16 units
    model.add(Dense(16, kernel_initializer='normal', activation='relu'))
    # Output layer with a binary classification (Democrat or Republican political party)
    model.add(Dense(1, kernel_initializer='normal', activation='sigmoid'))
    # Compile model
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    return model

from tensorflow.keras.wrappers.scikit_learn import KerasClassifier

# Wrap our Keras model in an estimator compatible with scikit_learn
estimator = KerasClassifier(build_fn=create_model, epochs=100, verbose=0)
# Now we can use scikit_learn's cross_val_score to evaluate this model identically to the others
cv_scores = cross_val_score(estimator, all_features, all_classes, cv=10)
cv_scores.mean()

0.9438405811786652

94% without even trying too hard! Did you do better? Maybe more neurons, more layers, or Dropout layers would help even more.